redis-statsd.py

```diff
@@ -102,56 +105,70 @@ def linesplit(socket):
-                )
-
-        for kg in KEYSPACE_GAUGES:
-            if kg in stats['keyspaces'][ks]:
-                send_metric('{}.keyspace.{}'.format(
-                    args.prefix, kg), 'g',
-                    float(stats['keyspaces'][ks][kg]), ['keyspace={}'.format(ks)]
-                )
-
-    out_sock.close()
-    time.sleep(10)
+    try:
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        (redis_host, redis_port) = args.redis_host.split(':')
+        s.connect((redis_host, int(redis_port)))
+        s.send("INFO\n")
+
+        stats = {}
+        stats['keyspaces'] = {}
+
+        for line in linesplit(s):
+            if '# Clients' in line:
+                for l in line.split("\n"):
+                    if ':keys' in l:
+                        (keyspace, kstats) = l.split(':')
+                        if keyspace not in stats['keyspaces']:
+                            stats['keyspaces'][keyspace] = {}
+                        for ks in kstats.split(','):
+                            (n, v) = ks.split('=')
+                            stats['keyspaces'][keyspace][n] = v.rstrip()
+
+                    elif ':' in l:
+                        (name, value) = l.split(':')
+                        stats[name] = value.rstrip()
+
+        s.close()
+
+        out_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+
+        for g in GAUGES:
+            if g in stats:
+                send_metric('{}.{}'.format(args.prefix, g), 'g', float(stats[g]))
+                logger.info('Metric sent: {}.{}'.format(args.prefix, g), 'g', float(stats[g]))
+
+        for c in COUNTERS:
+            if c in stats:
+                send_metric('{}.{}'.format(args.prefix, c), 'c', float(stats[c]))
+                logger.info('Metric sent: {}.{}'.format(args.prefix, c), 'c', float(stats[c]))
+
+        for ks in stats['keyspaces']:
+            for kc in KEYSPACE_COUNTERS:
+                if kc in stats['keyspaces'][ks]:
+                    send_metric('{}.keyspace.{}'.format(
+                        args.prefix, kc), 'c',
+                    float(stats['keyspaces'][ks][kc]), ['keyspace={}'.format(ks)]
+                    )
+                    logger.info('Metric sent: {}.keyspace.{}'.format(
+                        args.prefix, kc), 'c',
+                    float(stats['keyspaces'][ks][kc]), ['keyspace={}'.format(ks)]
+                    )
+
+            for kg in KEYSPACE_GAUGES:
+                if kg in stats['keyspaces'][ks]:
+                    send_metric('{}.keyspace.{}'.format(
+                        args.prefix, kg), 'g',
+                        float(stats['keyspaces'][ks][kg]), ['keyspace={}'.format(ks)]
+                    )
+                    logger.info('Metric sent: {}.keyspace.{}'.format(
+                        args.prefix, kg), 'g',
+                        float(stats['keyspaces'][ks][kg]), ['keyspace={}'.format(ks)]
+                    )
+
+        out_sock.close()
```

Comment:
Should probably define out_sock outside of `try` block and close in `finally` block to ensure we also close a possible open connection.